lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -0,0 +1,254 @@
+unnecessary.
+
+When specifying user permissions for this relation, a client application must send this charm valid
+JSON containing the following fields:
+
+{
+    "roles": ["A list of default opensearch roles to apply"],
+    "permissions": ["A list of default opensearch permissions to apply"],
+    "action_groups": ["A list of default opensearch action groups to apply"],
+}
+
+A role will be created for the relation with the permissions and action groups applied, and these
+roles will be mapped to a dedicated user for the relation, which will be removed with the relation.
+Default security values can be found in the opensearch documentation here:
+https://opensearch.org/docs/latest/security/access-control/index/.
+
+TODO add databag reference information
+TODO add tls
+"""
+
+import json
+import logging
+from typing import Dict, List
+
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseProvides,
+    DatabaseRequestedEvent,
+)
+from charms.opensearch.v0.constants_charm import (
+    ClientRelationBadRoleRequestMessage,
+    ClientRelationName,
+    PeerRelationName,
+)
+from charms.opensearch.v0.helper_databag import Scope
+from charms.opensearch.v0.helper_networking import units_ips
+from charms.opensearch.v0.helper_security import generate_hashed_password
+from charms.opensearch.v0.opensearch_users import OpenSearchUserMgmtError
+from ops.charm import CharmBase, RelationBrokenEvent, RelationDepartedEvent
+from ops.framework import Object
+from ops.model import BlockedStatus, Relation
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchProvider(Object):
+    """Defines functionality for the 'provides' side of the 'opensearch-client' relation.
+
+    Hook events observed:
+        - database-requested
+        - relation-departed
+        - relation-broken
+    """
+
+    def __init__(self, charm: CharmBase) -> None:
+        """Constructor for OpenSearchProvider object.
+
+        Args:
+            charm: the charm for which this relation is provided
+        """
+        super().__init__(charm, ClientRelationName)
+
+        self.charm = charm
+        self.unit = self.charm.unit
+        self.app = self.charm.app
+        self.opensearch = self.charm.opensearch
+        self.user_manager = self.charm.user_manager
+
+        self.relation_name = ClientRelationName
+        self.database_provides = DatabaseProvides(self.charm, relation_name=self.relation_name)
+
+        self.framework.observe(
+            self.database_provides.on.database_requested, self._on_database_requested
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_departed, self._on_relation_departed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+    def _relation_username(self, relation: Relation) -> str:
+        return f"{self.relation_name}_{relation.id}_user"
```

Comment:
I don't think we need postfix `_user` for every user. for example, nobody create accounts like `carlos_user@gmail.com`